Introduce tests for IPv6, update documentation about testing/contributing

hoegaarden opened the pull request:
Introduces some initial tests for IPv6 to run locally and on CircleCI

Implements part of: #182

hoegaarden commented:
I believe the failing test is a flake, however I am not allowed to rerun it.

pmichali commented:
There are additional issues, not being addressed... ensure-network() calls get-subnet(), which does not correctly parse the docker network, when in IPv6 mode. There will be two subnets from the docker inspect (one IPv4, one IPv6, in any order). 

ensure-network() does not set the GW for IPv6 mode correctly.

Nit: why do an inspection of the docker network, every time an I address is needed on the DinD subnet? Why not cache the value and use it throughout the script, like what was originally there (only programmable)?  IOW, instead of using a fixed default and allowing config override, like the original, calculate an available IPv4 subnet and store it in DIND_NET and DIND_SUBNET_SIZE, again allowing override via config, if desired. Use that subnet prefix everywhere, when creating IPs, like the original code.  I don't see the need to dynamically determine the subnet.

pmichali commented:
I'm assuming you are disabling IPv6 for multi-cluster, because each cluster needs a different DinD subnet?  I think it is OK to put a quick test at the start of the script to exit, if v6 and multi-cluster for now.

For IPv6 and multicluster, if you can introduce an ID number for each cluster (or generate a random one), it could be incorporated into the IP address easily... For example, instead of the default fd00:10::/64, it could be fd00:10:X::/64, where x=1..ffff.

I'm playing with modifying the current code to work with IPv6 for single cluster. If I get something, I can post diffs.

hoegaarden commented:
- We added one test to run on TravisCI which tests internal IPv6 connectivity & external via NAT64 (`test-case-ipv6-aaaa-disallowed`).
- For direct external IPv6 connectivity (`test-case-ipv6-aaaa-allowed`) we'd need to have a hosted CI with IPv6 enabled. Neither TravisCI nor CIrcleCI seem to offer that -- in fact, we couldn't find any hosted CI which supports IPv6.
- We added a section in the README describing how to run certain tests locally and which CI systems are in place.

We will continue to look into having a proper IPv6 CI, but this is at least a start.

@pmichali 
- Do you think the tests are OK for now or are we missing something important that could be tested even in a non -IPv6 environment?
- As soon as #192 is merged, we will rebase onto that.

Addresses parts of: #182

pmichali commented on README.md:
````diff
@@ -306,6 +308,39 @@ currently have some issues. You may still try running them though:
+$ TEST_CASE=ipv6-aaaa-disallowed ./test.sh
+# needs IPv6 enabled environment to succeed
+$ TEST_CASE=ipv6-aaaa-allowed ./test.sh
+```
````
FYI,  One should be able to enable AAAA mode, and even though they will be unable to ping an IPv6 capable external system given their network setup, and then do nslookup on pod and see that some external sites (e.g. google.com) should come up with an IPv6 address that does not have the DNS64 prefix, and others (e.g. github.com) should have the prefix.

May be a way to test both w/o running multiple tests.

pmichali commented on dind-cluster.sh:
```diff
@@ -265,6 +265,13 @@ fi
 DEFAULT_DIND_LABEL='mirantis.kubeadm_dind_cluster_runtime'
 : "${DIND_LABEL:=${DEFAULT_DIND_LABEL}}"
 
+# TODO -- make multi cluster && ipv6 work
```
This is in PR 192 (which I hope upstreams VERY soon).

pmichali commented on test.sh:
```diff
@@ -308,6 +308,120 @@ function fail() {
+}
+
+function setup-ipv6-test() {
+  # Currently only testing with a specific version of
```
Ideally we want to test with 1.9+ versions

hoegaarden replied:
Sure. I want to get the test case(s) in place first. Having them run for multiple versions should then be a rel. easy addition.

pmichali commented on test.sh:
```diff
@@ -308,6 +308,120 @@ function fail() {
+
+  pod_on_node1_ip="$( get-ip-for-pod 'pod-on-node1' )"
+
+  $k exec -ti pod-on-node1 -- ping6 -c1 $ipv6_enabled_host || {
```
Mentioned above, it may be easier to do nslookup on the pod for google.com and github.com and check that the former does NOT start with the DNS64 prefix and the latter does start with it.

pmichali commented on test.sh:
```diff
@@ -308,6 +308,120 @@ function fail() {
+
+# with DIND_ALLOW_AAAA_USE set
+function test-case-ipv6-aaaa-allowed() {
+  export DIND_ALLOW_AAAA_USE=some-value
```
use true, instead of some-value.

pmichali commented on test.sh:
```diff
@@ -308,6 +308,120 @@ function fail() {
+  local ipv6_host='ds.test-ipv6.noroutetohost.net'
+  local ipv6_host_ns='8.8.8.8'
+
+  expected_ipv6_addr="$( dig +short "@${ipv6_host_ns}" "${ipv6_host}" aaaa )"
```
Although it may be nice to actually ping to an IPV6 destination, I'm wondering if that will be problematic, as some sites, like google.com have multiple IPv6 addresses.  A compromise may be to just do a ping (by name) to an IPv4 site, and ensure that it works (and that NAT64 is working), and then use the nslookup mentioned before to verify that DNS64 is working for both AAAA and non-AAAA destinations.

pmichali commented on test.sh:
```diff
@@ -308,6 +308,120 @@ function fail() {
+  ping-tests
+
+  IP_MODE='ipv6' \
+    bash -x ./dind-cluster.sh clean
```
Would be good to do a down, verify that tayga and bind9 containers are still running, do an up, and make sure that ping test works again, and then do the clean and make sure tayga and bind9 containers are NOT running.

pmichali commented:
See PR #192, which has merged, as some of the code changes in dind-cluster.sh are handled.  Also, please see Issue #182 for ideas on IPv6 testing for CI.

hoegaarden commented:
@pmichali PTAL

We are aware, that we don't test every aspect of your list yet, but we'd like to get that out sooner then later and maybe add the additional  cases from #182 and potentially others one by one.

@mariantalla

hoegaarden commented:
@pmichali -- question regarding https://github.com/kubernetes-sigs/kubeadm-dind-cluster/pull/190#discussion_r209030154 :

> Ideally we want to test with 1.9+ versions

We saw that `kube-dashboard` and `kube-dns` pods were failing for `v1.8` after the snapshotting `k-d-c` does. We didn't investigate too much yet, partially also because you seem to be OK with not supporting 1.8.

Could you please add some context about IPv6 and 1.8: Is it not (well) supported? Is it that we just don't care about more the `n-2`?

pmichali commented on README.md:
```diff
@@ -322,6 +322,51 @@ currently have some issues. You may still try running them though:
+- [TravisCI](https://travis-ci.org/kubernetes-sigs/kubeadm-dind-cluster)
+- [CircleCI](https://circleci.com/gh/kubernetes-sigs/kubeadm-dind-cluster)
+
+To run tests locally, find the test function you wish to run (it is called
```
Can you check with @ivan4th? I thought he mentioned that test.sh was being deprecated and that we should be doing CircleCI tests only.

pmichali replied:
Did you get to touch base with Ivan?

pmichali commented on README.md:
```diff
@@ -322,6 +322,51 @@ currently have some issues. You may still try running them though:
+those host does not get routed through NAT64. In that case the host running the
+outer docker daemon needs to have external IPv6 available.
+
+As our CI currently does not allow for external IPv6 the external ping tests
```
We could, for this case, just check that name lookup works correctly. Namely, that for external sites that support IPv6, the address should NOT have the DNS64 prefix and external sites that only support IPv4 will always have the DNS64 prefix (and use an embedded IPv4 address). I wouldn't advocate checking the IPv4 address, but only whether or not the prefix exists, in these too cases.

hoegaarden replied:
It seems that this is what the tests are doing at the moment, if I've got this right. Are there points that you'd rather see implemented differently?

pmichali commented on test.sh:
```diff
@@ -308,6 +308,237 @@ function fail() {
+}
+
+function ping-tests-external() {
+  if [ -n "${DIND_ALLOW_AAAA_USE:-}" ]
```
mentioned above, could just check name lookup for IPv6 capable external hosts, but actually do a ping to IPv4 hosts, rather than exist the test, if set.

hoegaarden replied:
Is this covered by [this](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/blob/870ba8c4a5c5bf0289a97ef78710f38f2295497e/test.sh#L501-L526) ?

pmichali commented on test.sh:
```diff
@@ -308,6 +308,237 @@ function fail() {
+
+  test -n "$ns" && ns="@${ns}"
+
+  $kubectl exec -ti "$pod" -- dig +short "${name}" "${type}" ${ns} | sed 's/\r//g'
```
For pods, should just be able to use "ip a" or ifconfig for eth0. I wouldn't think there would be an A record for the pods. should only have IPv6 address.

hoegaarden replied:
Same as [below](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/pull/190#discussion_r210202524): This is a bad name, it actually tests DNS resolution from within the pod.

pmichali commented on test.sh:
```diff
@@ -308,6 +308,237 @@ function fail() {
+
+  for target in "${ipv6_enabled_host}" "${ipv4_only_host}"
+  do
+    aaaa_from_node="$( get-addr-on-node 'kube-node-1' "${target}" AAAA )" || {
```
Not sure of the usefulness of getting the IPs for the nodes. Goal is to ensure pod to pod access.

hoegaarden replied:
pod to pod (ping) is tested [here](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/pull/190/files#diff-1b531c5e405e0bc95890d52cf2f2f5b5R436). This is testing the DNS resolution of a given name on a pod.

pmichali commented on test.sh:
```diff
@@ -308,6 +308,237 @@ function fail() {
+    aaaa_from_pod="$( get-addr-on-pod 'pod-on-node1' "${target}" AAAA )" || {
+      fail "Expected to successfully resolve ${target}'s IPv6 address on a pod"
+    }
+    a_from_pod="$( get-addr-on-pod 'pod-on-node1' "${target}" A )" || {
```
I would think that the pods should not have an IPv4 address, and only should have an IPv6 address.

hoegaarden replied:
I see the confusion: `get-addr-on-pod` functions are not about getting the address of a pod, it is about resolving a name from within the pod. We can rename the function(s) to make their purpose clearer.

pmichali replied:
Maybe break the logic out into a function that indicates it is testing name lookup and call it, like you are doing in L529-530 with the ping tests, so that test-case-ipv6 is at a higher level, calling 3 tests, internal ping, external ping, and name lookup. Later more can be added.

pmichali commented:
@hoegaarden IPv6 support was added in K8s 1.9 (as alpha feature), so it is not supported at all in 1.8 and older.

pmichali commented on test.sh:
```diff
@@ -308,6 +308,237 @@ function fail() {
+    aaaa_from_node="$( get-addr-on-node 'kube-node-1' "${target}" AAAA )" || {
+      fail "Expected to successfully resolve ${target}'s IPv6 address on kube-node-1"
+    }
+    a_from_node="$( get-addr-on-node 'kube-node-1' "${target}" A )" || {
```
Since this is IPv6 only mode, we won't ever be accessing hosts by IPv4 addresses, so tests like L461-463, L468-L470, L475-477,... are not really needed.

pmichali commented on README.md:
```diff
@@ -322,6 +322,51 @@ currently have some issues. You may still try running them though:
+[TravisCI](https://travis-ci.org/kubernetes-sigs/kubeadm-dind-cluster). There
+are two slightly different kind of tests which run for all major version starting from `v1.9`:
+
+#### `TEST_CASE=ipv6 ./test.sh`
```
Consider calling the test case "ipv6-only", as in the future, we'll have dual-stack support support in K8s and maybe it'll reduce confusion?

hoegaarden commented:
@pmichali -- *PTAL* and let us know if that test setup looks good to you

- We refactored the test a bit based on your comments
- We dropped Ivan a message about Travis vs. CircleCI vs. `./test.sh` -- we will let you know once we know where this is going

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+  export DIND_IMAGE="mirantis/kubeadm-dind-cluster:${v}"
+  docker pull "${DIND_IMAGE}"
+
+  # Globally exported CIDR not applicable for IPv6
```
Don't need to clear this, as dind-cluster.sh will set to default, if not specified by user.

hoegaarden replied:
This is an artifact from having those tests in `./test.sh` ... I will remove that.

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+  # Add the directory we will download kubectl into to the PATH
+  export PATH="${PATH}:${KUBECTL_DIR}"
+
+  # Explicitiely set the DNS64 prefix
```
This is defaulted too, in dind-cluster.sh

hoegaarden replied:
This is explicitly set, so in the tests we can check if a record is prefixed with `DNS64_PREFIX`. If we'd leave it to `./dind-cluster.sh` to default that, we couldn't really use that in the test assertions. I'll try yo add a better comment describing that.

pmichali replied:
Gotcha, was thinking the setting from dind-cluster.sh was visible.

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+  export ipv6_enabled_host_ns='8.8.8.8'
+
+  # kubectl setup
+  export kubectl="kubectl --context dind"
```
I'm pretty sure this is no longer needed either, as the context has been fixed for single cluster use.

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+  export k8s_img='tutum/dnsutils'
+}
+
+function lookup-address() {
```
would dns-lookup-ip-by-type be clearer?

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+  $kubectl exec -ti "$pod" -- dig +short "${name}" "${type}" ${ns} | sed 's/\r//g'
+}
+
+function lookup-address-on-node() {
```
lookup-ip-address-on-node or lookup-node-ip?  Just trying to reduce the ambiguity of "address" on these function names.

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+
+function ping-tests-external() {
+  $kubectl exec -ti pod-on-node1 -- ping6 -n -c1 $ipv6_enabled_host || {
+    fail "Expected ping6 to $ipv6_enabled_host to succeed"
```
Curious... if the test fails here and other checks, will the pod(s) be cleaned up?

hoegaarden replied:
Nope, when any test fails, no cleanup will be done and the test is aborted right away. This is, at least from my point of view, deliberate: In CI everything will be torn down anyway, locally this gives the user a chance to debug.

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+  dig +short ${name} ${type} ${ns} | sed 's/\r//g'
+}
+
+function lookup-address-on-pod() {
```
would dns-lookup-ip-on-pod-by-type be clearer?

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+  echo "${aaaa_from_host},${aaaa_from_k8s}"
+}
+
+function verify-lookup-with-dns64-only() {
```
verify-lookup-always-using-dns64-prefix?

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+  fi
+}
+
+function verify-lookup-with-ext-dns-for-ipv6-host() {
```
verify-lookup-using-ipv6-address-for-ipv6-hosts?

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+  fi
+}
+
+function verify-lookup-with-ext-dns-for-ipv4-only-host() {
```
verify-lookup-using-dns64-prefix-for-ipv4-hosts?

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+function test-case-ipv6-only() {
+  setup-ipv6-only-test
+
+  bash -x ./dind-cluster.sh up
```
Using -x for debug purposes?

hoegaarden replied:
`-x` because this was the default in `./test.sh` and I just went with it. I find it nice but I am happy to remove it to be more consistent with other CIrcleCI test or if people prefer to not use this.

pmichali replied:
It displays output that is useful during debug, but not sure we need all that output for normal runs. You could use a variable DEBUG="-x", so it could be turned on/off, if there are problems with running dind-cluster.sh.

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+  else
+    {
+      echo ''
+      echo 'WARNING: not testing direct external IPv6 connectivity.'
```
Maybe a NOTE vs WARNING?

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,244 @@
+    } >&2
+  fi
+
+  bash -x ./dind-cluster.sh clean
```
-x for debug? Pulling later?

hoegaarden replied:
regarding `-x` see https://github.com/kubernetes-sigs/kubeadm-dind-cluster/pull/190#discussion_r210905163

> Pulling later?

I am not sure what you mean by that? Could you please elaborate?

pmichali replied:
The -x just shows line by line execution of the script. Useful for debug, but I don't know if you need the option in the final run, especially since it is just teardown.

hoegaarden commented:
@pmichali *PTAL*

pmichali commented on README.md:
```diff
@@ -322,6 +322,89 @@ currently have some issues. You may still try running them though:
+which have a AAAA record, the IP address is used, traffic to those hosts does
+not get routed through NAT64. In that case the host running the outer docker
+daemon would need to have external IPv6 available to actually communicate with
+external IPv6 hosts. Therefore (becuase none of our CI systems can provide
```
nit: becuase -> because

hoegaarden commented:
@pmichali Thanks for your help on that!

@ivan4th / @pigmej Could you maybe also check the PR and give the *lgtm* if you are happy with it?

pmichali commented on test/test-ipv6-only.sh:
```diff
@@ -0,0 +1,248 @@
+}
+
+function ping-tests-external() {
+  kubectl exec -ti pod-on-node1 -- ping6 -n -c1 $ipv6_enabled_host || {
```
Just curious as to whether the ping test should be with a count of 5 vs 1 (in case there is some network glitch that causes one ping to fail)? I'm not sure what happens if several pass and a few fail, w.r.t return code.

hoegaarden replied:
I'd say we keep it as is (one ping). We had it running quite often now (locally, on Travis & on CircleCI) and one ping was always fine. If we see that those tests are getting flaky we can see if we want to use multiple pings.

ivan4th commented on README.md:
```diff
@@ -322,6 +322,89 @@ currently have some issues. You may still try running them though:
+#### TravisCI, `./test.sh` -- DEPRECATED
+
+There are some tests in `./test.sh`. These were the tests which ran on TravisCI
+when TravisCI was the default CI system.
```
Small correction, sorry, perhaps I was not precise enough when I was trying to describe my idea of using TravisCI. TravisCI is still supported for people who want to use kdc there, but only a few cases are run there now instead of the full suite. The corresponding unused parts of test.sh can be removed. So test.sh / TravisCI is not completely deprecated, just parts of it.

hoegaarden replied:
Got it -- will update the docs on that.

ivan4th commented on .travis.yml:
```diff
@@ -14,3 +14,6 @@ before_install:
   - if [ -n "${TWAIT:-}" ]; then travis_wait 35 ./test.sh; else ./test.sh; fi
+
+before_script:
+  - echo 0 | sudo tee /proc/sys/net/ipv6/conf/all/disable_ipv6
```
We don't test IPv6 on Travis, so why do we need this?

ivan4th commented on fixed/dind-cluster-stable.sh:
```diff
@@ -1152,9 +1152,9 @@ function dind::setup_external_access_on_host {
   fi
   local main_if=`ip route | grep default | awk '{print $5}'`
   local bridge_if=`ip route | grep ${NAT64_V4_SUBNET_PREFIX}.0.0 | awk '{print $3}'`
-  docker run --entrypoint /sbin/ip6tables --net=host --rm --privileged ${DIND_IMAGE} -t nat -A POSTROUTING -o $main_if -j MASQUERADE
```
`dind-cluster-stable.sh` is supposed to be used with `mirantis/kubeadm-dind-cluster:stable` image and is only updated when we need to update that stable image. We're not updating it right now, so perhaps we should leave `fixed/dind-cluster-stable.sh` untouched.

hoegaarden replied:
- that's a change done by `genfixed.sh`
- CI would fail if we didn't have that change
- This code was there before already as is, it has just been pushed into the function [`dind::ip6tables-on-hostnet`](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/blob/b91d8899a3bee39683f427b855fb34866f006fee/fixed/dind-cluster-stable.sh#L1190)
- `DIND_IMAGE` is set specifically to `mirantis/kubeadm-dind-cluster:stable` [here](https://github.com/kubernetes-sigs/kubeadm-dind-cluster/blob/b91d8899a3bee39683f427b855fb34866f006fee/fixed/dind-cluster-stable.sh#L52)

With all that, could you more context what's the issue with this change? Am I missing something?

pmichali replied:
I thought this change was just pulling out duplication and placing into a common function. It should be a refactoring and not altering the script functionality, right?

ivan4th commented:
/approve
/lgtm

ivan4th commented:
Thanks for the PR!
